src/backtrace.c

```diff
@@ -0,0 +1,682 @@
+ * 
+ */
+#include <stdint.h>
+#include <stdalign.h>
+#include <stdlib.h>
+#include <string.h>
+#include "backtrace.h"
+#include "backtrace_internal.h"
+#include "debug.h"
+#include "n64sys.h"
+#include "dma.h"
+#include "utils.h"
+#include "exception.h"
+#include "interrupt.h"
+#include "rompak_internal.h"
+
+/** @brief Enable to debug why a backtrace is wrong */
+#define BACKTRACE_DEBUG 0
+
+/** @brief Function alignment enfored by the compiler (-falign-functions). 
+ * 
+ * @note This must be kept in sync with n64.mk.
+ */
+#define FUNCTION_ALIGNMENT      32
+
+#define MAX_FILE_LEN 120        ///< Maximum length of a file name in a backtrace entry
+#define MAX_FUNC_LEN 120        ///< Maximum length of a function name in a backtrace entry
+
+/** 
+ * @brief Symbol table file header
+ * 
+ * The SYMT file is made of three main tables:
+ * 
+ * * Address table: this is a sequence of 32-bit integers, each representing an address in the ROM.
+ *   The table is sorted in ascending order to allow for binary search. Moreover, the lowest 2 bits
+ *   of each address can store additional information: If bit 0 is set to 1, the address is the start
+ *   of a function. If bit 1 is set to 1, the address is an inline duplicate. In fact, there might be
+ *   multiple symbols at the same address for inlined functions, so we need one entry in this table
+ *   for each entry; all of them will have the same address, and all but the last one will have bit
+ *   1 set to 1.
+ * * Symbol table: this is a sequence of symbol table entries, each representing a symbol. The size
+ *   of this table (in number of entries) is exactly the same as the address table. In fact, each
+ *   address of the address table can be thought of as an external member of this structure; it's
+ *   split externally to allow for efficiency reasons. Each entry stores the function name, 
+ *   the source file name and line number, and the binary offset of the symbol within the containing
+ *   function.
+ * * String table: This tables can be thought as a large buffer holding all the strings needed by all
+ *   symbol entries (function names and file names). Each symbol entry stores a string as an index
+ *   within the symbol table and a length. This allows to reuse the same string (or prefix thereof)
+ *   multiple times. Notice that strings are not null terminated in the string table.
+ * 
+ * The SYMT file is generated by the n64sym tool during the build process.
+ */
+typedef struct alignas(8) {
+    char head[4];           ///< Magic ID "SYMT"
+    uint32_t version;       ///< Version of the symbol table
+    uint32_t addrtab_off;   ///< Offset of the address table in the file
+    uint32_t addrtab_size;  ///< Size of the address table in the file (number of entries)
+    uint32_t symtab_off;    ///< Offset of the symbol table in the file
+    uint32_t symtab_size;   ///< Size of the symbol table in the file (number of entries); always equal to addrtab_size.
```

Comment:
Is this for "just in case"?

Reply:
Yeah, I figured it out that keeping the header more generic wasn't going to be a problem.

Reply:
On second thought what about "reserving" a few bytes in the header as well? OTOH it is a matter of a rebuild but that extra length made me think about this as well

Reply:
It's actually not necessary because anything after the header is accessed via offsets specified in the header itself. That's actually by design.

Reply:
I mean in case we need to store anything else in the header in the future. I didn't get your point?

Reply:
Hmm I think I get it now, we will simply shift everything if we need any space.